frontend/z3_types.py

```diff
@@ -0,0 +1,232 @@
+"""The type-system for Python 3 encoded in Z3.
+
+Limitations:
+    - The tuples have a maximum size of 5 elements.
+    - Function calls can have no more than 5 arguments.
+    - Multiple inheritance is not supported.
+    - Functions with generic type variables are not supported.
+"""
+from z3 import *
+
+# ----------- Declare the types data-type -----------
+
+type_sort = Datatype("Type")
+type_sort.declare("object")
+
+type_sort.declare("none")
+
+type_sort.declare("number")
+type_sort.declare("complex")
+type_sort.declare("float")
+type_sort.declare("int")
+type_sort.declare("bool")
+
+type_sort.declare("sequence")
+type_sort.declare("str")
+type_sort.declare("bytes")
+type_sort.declare("list", ("list_type", type_sort))
+
+type_sort.declare("set", ("set_type", type_sort))
+type_sort.declare("dict", ("dict_key_type", type_sort), ("dict_value_type", type_sort))
+
+type_sort.declare("tuple")
+type_sort.declare("tuple_1", ("tuple_1_arg_1", type_sort))
+type_sort.declare("tuple_2", ("tuple_2_arg_1", type_sort), ("tuple_2_arg_2", type_sort))
+type_sort.declare("tuple_3", ("tuple_3_arg_1", type_sort), ("tuple_3_arg_2", type_sort), ("tuple_3_arg_3", type_sort))
+type_sort.declare("tuple_4", ("tuple_4_arg_1", type_sort), ("tuple_4_arg_2", type_sort), ("tuple_4_arg_3", type_sort),
+                  ("tuple_4_arg_4", type_sort))
+type_sort.declare("tuple_5", ("tuple_5_arg_1", type_sort), ("tuple_5_arg_2", type_sort), ("tuple_5_arg_3", type_sort),
+                  ("tuple_5_arg_4", type_sort), ("tuple_5_arg_5", type_sort))
+
+type_sort.declare("func")
+type_sort.declare("func_0", ("func_0_return", type_sort))
```

Comment:
I'm not sure I understand what these are for. I would have expected that we have different ones for each function in the program or something?